### PR TITLE
Add prop to Table header render function

### PR DESCRIPTION
**Pull request by msnyder-msft**
This PR is intended to address the comments in #29264.
It allows consumers of the Table component to pass in parameters of any type to the `renderHeaderCell` function. There are multiple potential scenarios that justify enabling this such as conditional styling, or dynamic column headers, etc.

The reason for using the `unknown` type is that there was no good way that we found to use a generic type (similar to how `renderCell` is shaped) without causing both a breaking change and/or forcing users to define 2 generics for the row when they really only cared about either the header or the cell data type. Ex: `<DataGridRow<unknown, ColumnData>>` for the data rows or `<DataGridRow<HeaderData, unknown>>` for the header rows.

CC: @ling1726

### CODE REVIEW

File: change/@fluentui-react-table-2e99fd0b-42bc-411d-9c1c-353f8cbbee19.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add optional parameter to Table column renderHeaderCell callback",
```
Comment (ling1726) [inline change/@fluentui-react-table-2e99fd0b-42bc-411d-9c1c-353f8cbbee19.json:3]:
```suggestion
  "comment": "feat; Add optional parameter to Table column renderHeaderCell callback",
```

**Review by ling1726 — APPROVED** [review]

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Avatar-mount_base.html">611</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Avatar-mount.html">614</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Button-mount_base.html">308</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Button-mount.html">320</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Field-mount_base.html">1138</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Field-mount.html">1112</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProvider-mount_base.html">682</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProvider-mount.html">704</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-mount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-mount.html">82</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">64</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">63</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">74</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">71</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/InfoButton-mount_base.html">14</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/InfoButton-mount.html">10</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/MakeStyles-mount_base.html">845</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/MakeStyles-mount.html">871</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Persona-mount_base.html">1729</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/Persona-mount.html">1729</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/SpinButton-mount_base.html">1359</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29285/perf-test-react-components/SpinButton-mount.html">1350</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr>              |  `69.227 kB`<br />`19.584 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `206.783 kB`<br />`59.085 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `40.793 kB`<br />`13.521 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |     `6.48 kB`<br />`2.203 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/DataGrid.fixture.js'>DataGrid</abbr>                                                                                 |  `155.92 kB`<br />`43.463 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/PrimitivesOnly.fixture.js'>Table (Primitives only)</abbr>                                                            |  `42.715 kB`<br />`13.341 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableAsDataGrid.fixture.js'>Table as DataGrid</abbr>                                                                 |  `129.19 kB`<br />`34.771 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSelectionOnly.fixture.js'>Table (Selection only)</abbr>                                                         |  `74.557 kB`<br />`20.071 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSortOnly.fixture.js'>Table (Sort only)</abbr>                                                                   |   `73.164 kB`<br />`19.67 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/bea2e1d48616fe1b48c72b119bcf79dca60ff44a'>bea2e1d48616fe1b48c72b119bcf79dca60ff44a</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Pull request merged**